presto-iceberg/src/main/java/com/facebook/presto/iceberg/function/changelog/ApplyChangelogFunction.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.iceberg.function.changelog;
+
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.block.BlockBuilder;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.spi.function.AggregationFunction;
+import com.facebook.presto.spi.function.AggregationState;
+import com.facebook.presto.spi.function.BlockIndex;
+import com.facebook.presto.spi.function.BlockPosition;
+import com.facebook.presto.spi.function.CombineFunction;
+import com.facebook.presto.spi.function.InputFunction;
+import com.facebook.presto.spi.function.OutputFunction;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.function.TypeParameter;
+import io.airlift.slice.Slice;
+import org.apache.iceberg.ChangelogOperation;
+
+import java.util.Optional;
+
+import static com.facebook.presto.spi.function.SqlFunctionVisibility.HIDDEN;
+
+@AggregationFunction(value = "apply_changelog", visibility = HIDDEN)
+public class ApplyChangelogFunction
+{
+    public static final String NAME = "apply_changelog";
+
+    private ApplyChangelogFunction() {}
+
+    @InputFunction
+    @TypeParameter("T")
+    public static void input(
+            @TypeParameter("T") Type type,
+            @AggregationState ApplyChangelogState state,
+            @SqlType("bigint") long ordinal,
+            @SqlType("varchar") Slice operation,
+            @BlockPosition @SqlType("T") Block block,
+            @BlockIndex int index)
+    {
+        ChangelogRecord record = state.get();
+        if (record == null) {
+            record = new ChangelogRecord(type);
+        }
+        record.add((int) ordinal, operation, block.getSingleValueBlock(index));
+        state.set(record);
+    }
+
+    @CombineFunction
+    public static void combine(@TypeParameter("T") Type type,
+            @AggregationState ApplyChangelogState state,
+            @AggregationState ApplyChangelogState otherState)
+    {
+        ChangelogRecord r = Optional.ofNullable(state.get())
+                .map(actual -> Optional.ofNullable(otherState.get()).map(actual::merge).orElse(actual)).orElse(otherState.get());
```

Comment:
I find this nested optional block to be confusing.  I'm wondering if simple if/else might be more LOCs yet more readable.

Reply:
I agree, the nested optionals can be confusing at first, but using them prevents NPEs. The `if`-condition version of the function would look like:

```java
        ChangelogRecord record;
        if (state.getRecord() == null && otherState.getRecord() == null) {
            return;
        } else if (state.getRecord() == null && otherState.getRecord() != null) {
            record = otherState.getRecord();
        } else if (state.getRecord() != null && otherState.getRecord() == null) {
            record = state.getRecord();
        } else {
            record = state.getRecord().merge(otherState.getRecord());
        }
        state.setRecord(record);
```